solutions/python-igraph/clique_percolation/clique_percolation.py

```diff
@@ -0,0 +1,88 @@
+import igraph as ig
+import matplotlib.pyplot as plt
+
+EXAMPLE_K = 3
+
+def k_cliques(g, k):
+    """Returns a list of all cliques of size k in Graph g. Each clique is given as a set of vertex ids in g.
+    """
+    return g.cliques(k, k)
+
+def k_communities(g, k, min_common_vertices):
+    """
+    Given a graph and size k, return a list of communities found through clique
+    percolation with at least min_common_vertices shared.
+    A community is the connected component of cliques where two cliques overlap
+    in at least k-1 vertices.
+    Each community is a set of vertices in that community
+
+    Approach: generate a new clique graph where each k-clique is a vertex, and 
+    connect vertices if the cliques share min_common_vertices
+    """
+    if (min_common_vertices > k):
+        raise ValueError("Error: min_common_vertices=%d must be greater than k=%d" % (min_common_vertices, k))
+
+    # Each clique can be identified by it's index in 'cliques'
+    cliques = [set(x) for x in k_cliques(g, k)]
+    num_cliques = len(cliques)
+    # Graph storing each clique as a vertex
+    clique_graph = ig.Graph()
+    clique_graph.add_vertices(num_cliques)
+    cg_edges = []
+    # List of communities as a list of list of clique indices
+    clique_communities = []
+    # List of communities as a list of list of set of vertices
+    communities = []
+
+    # Get connected cliques
+    for c1 in range(num_cliques):
+        for c2 in range(c1, num_cliques):
+            # If the two cliques share at least min_common_vertices,
+            # they are part of the same community
+            if c1 != c2 and len(cliques[c1].intersection(cliques[c2])) >= min_common_vertices:
+                cg_edges.append((c1, c2))
+    
+    # Add edges for clique graph
+    clique_graph.add_edges(cg_edges)
+
+    # Merge cliques accordingly into communities
+    to_merge = list(range(num_cliques))
+    while (len(to_merge) > 0):
+        # Get first unmerged vertex
+        idx = to_merge[0]
+        # Get all connected vertices
+        vertices = clique_graph.subcomponent(idx, mode="all")
+        # Remove all connected vertices
+        for v in vertices:
+            to_merge.remove(v)
+        clique_communities.append(vertices)
```

Comment:
I think that instead of this while loop, you can simply call `clique_graph.clusters()`, which gives you a `VertexClustering` object that has all that you need.

Reply:
Basically you will need to iterate over the `VertexClustering` to get the connected components of the clique graph.